sqld/src/hrana/ws/handshake.rs

```diff
@@ -14,14 +19,29 @@ pub enum WebSocket {
     Upgraded(tokio_tungstenite::WebSocketStream<hyper::upgrade::Upgraded>),
 }
 
-pub async fn handshake_tcp(socket: tokio::net::TcpStream) -> Result<(WebSocket, Version)> {
+fn extract_namespace<B>(req: &Request<B>) -> anyhow::Result<Bytes> {
+    let host = req.headers().get("host").context("missing host header")?;
+    Ok(split_namespace(std::str::from_utf8(host.as_bytes())?)?)
+}
+
+pub async fn handshake_tcp(
+    socket: tokio::net::TcpStream,
+    allow_default_ns: bool,
+) -> Result<(WebSocket, Version, Bytes)> {
     let mut version = None;
+    let mut namespace = None;
     let callback = |req: &http::Request<()>, resp: http::Response<()>| {
         let (mut resp_parts, _) = resp.into_parts();
         resp_parts
             .headers
             .insert("server", http::HeaderValue::from_static("sqld-hrana-tcp"));
 
+        namespace = match extract_namespace(req) {
+            Ok(ns) => Some(ns),
+            Err(_) if allow_default_ns => Some(DEFAULT_NAMESPACE_NAME.into()),
```

Comment:
I don't think you can do that since `namespace` is set by the callback body, and Rust can't prove that the callback will be invoked at any point, leaving the namespace uninitialized. Or maybe I misunderstood your suggestion?